source/cipher.c

```diff
@@ -357,7 +357,10 @@ static int aws_cryptosdk_private_derive_key_v2(
     const struct aws_byte_buf myikm  = aws_byte_buf_from_array(data_key->keybuf, props->data_key_len);
 
     memset(content_key->keybuf, 0, sizeof(content_key->keybuf));
-    memset(commitment->buffer, 0, props->commitment_len);
+    if (props->commitment_len) {
+        assert(AWS_MEM_IS_WRITABLE(commitment->buffer, props->commitment_len));
```

Comment:
In this case,
```c
#    define AWS_MEM_IS_WRITABLE(base, len) (((len) == 0) || (base))
```
So if `props->commitment_len` is 0, we can still pass a NULL pointer to `memset`. I'd recommend the following:
```c
assert(commitment->buffer && AWS_MEM_IS_WRITABLE(commitment->buffer, props->commitment_len));
```

Reply:
But if `props->commitment_len` is 0, neither `assert` or `memset` are executed. Right?

In this case, `AWS_MEM_IS_WRITABLE(base, len)` is equivalent to `base`, so we can just write:
```
assert(commitment->buffer);
``` 

Thanks for bringing this up!

Reply:
Thanks all. I agree with Adrian here - the guard ensures that the assert is only reached when `props->commitment_len` is positive.

However, if I understand correctly, `AWS_MEM_IS_WRITABLE(base, len)` might not be equivalent to `base` when running under CBMC:

```c
/**
 * Define function contracts.
 * When the code is being verified using CBMC these contracts are formally verified;
 * When the code is built in debug mode, they are checked as much as possible using assertions
 * When the code is built in production mode, non-fatal contracts are not checked.
 * Violations of the function contracts are undefined behaviour.
 */
#ifdef CBMC
// snip
#    define AWS_MEM_IS_READABLE(base, len) (((len) == 0) || (__CPROVER_r_ok((base), (len))))
#    define AWS_MEM_IS_WRITABLE(base, len) (((len) == 0) || (__CPROVER_r_ok((base), (len))))
#else
// snip
/**
 * The C runtime does not give a way to check these properties,
 * but we can at least check that the pointer is valid. */
#    define AWS_MEM_IS_READABLE(base, len) (((len) == 0) || (base))
#    define AWS_MEM_IS_WRITABLE(base, len) (((len) == 0) || (base))
#endif /* CBMC */
```

(from https://github.com/awslabs/aws-c-common/blob/a61f2ea5fbc9ac2eee9fe32f7a3fdbd399d9fbdb/include/aws/common/assert.h#L73)

Can I safely simplify the `AWS_MEM_IS_WRITABLE(base, len)` call to just `base` in the both cases?

Reply:
You're both right. I believe in this case, the best option would be
```c
assert(commitment->buffer);
```

Reply:
Got it, I'll change that then.